File: ResponsibilityChain/ResponsibilityChainPattern.java
class Request {
}

abstract class Handler {
    private Handler nextHandler;

    public abstract void processRequest(Request request);

    public void setNextHandler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    public Handler getNextHandler() {
        return nextHandler;
    }
}

class ConcreteHandler1 extends Handler {
    @Override
    public void processRequest(Request request) {
        System.out.println("ConcreteHandler1: "
             + "I am processing request: " + request);
        if (getNextHandler() != null)
            getNextHandler().processRequest(request);
    }
}

class ConcreteHandler2 extends Handler {
    @Override
    public void processRequest(Request request) {
        System.out.println("ConcreteHandler2: "
             + "I am processing request: " + request);
        if (getNextHandler() != null)
            getNextHandler().processRequest(request);
    }
}

public class ResponsibilityChainPattern {
    public static void main(String[] args) {
        Handler handler1 = new ConcreteHandler1();
        Handler handler2 = new ConcreteHandler2();
        handler1.setNextHandler(handler2);

        handler1.processRequest(new Request());
    }
}

File: ResponsibilityChain/ResponsibilityChainPattern.py
# coding: utf8

import abc


class Handler(object):
    __metaclass__ = abc.ABCMeta

    def __init__(self, next_handler=None):
        self._next_handler = next_handler

    def handle(self, request):
        self._handle(request)
        if self.next_handler is not None:
            self.next_handler.handle(request)

    @abc.abstractmethod
    def _handle(self, request):
        pass

    @property
    def next_handler(self):
        return self._next_handler


class ConcreteHandler1(Handler):
    def _handle(self, request):
        print("handle request in ConcreteHandler1")


class ConcreteHandler2(Handler):
    def _handle(self, request):
        print("handle request in ConcreteHandler2")


if __name__ == "__main__":
    import unittest


    class ResponsibilityChainTest(unittest.TestCase):
        def testResponsibilityChain(self):
            class Request(object):
                pass

            handler2 = ConcreteHandler2()
            handler1 = ConcreteHandler1(handler2)
            handler1.handle(Request())


    unittest.main()
